src/gt4py/next/ffront/foast_to_itir.py

```diff
@@ -318,12 +318,22 @@ def visit_Call(self, node: foast.Call, **kwargs) -> itir.Expr:
     def _visit_astype(self, node: foast.Call, **kwargs) -> itir.FunCall:
         assert len(node.args) == 2 and isinstance(node.args[1], foast.Name)
         obj, dtype = node.args[0], node.args[1].id
-
-        # TODO check that we test astype that results in a itir.map_ operation
-        return self._map(
-            im.lambda_("it")(im.call("cast_")("it", str(dtype))),
-            obj,
-        )
+        if isinstance(obj, foast.TupleExpr):
+            casted_elements = []
```

Comment:
This will fail for nested tuples. I would suggest to first extend it (using some recursive function) and then try to generalize it to a function that is independent of the `cast` use-case here (analogous to `apply_to_primitive_constituents`). Let's start with a first draft and look at it together again, not sure how well this works out.

Reply:
The first suggestion (recursive function not generalized) has been added and here a suggestion for the generalized function analogous to `apply_to_primitive_constituents` which has proven to be very useful: 
```python

def process_elements(obj, process_func, **kwargs):
    """
    Example:
    result = process_elements(obj, lambda x: im.call("cast_")(x, str(dtype)), **kwargs)

    """
    if isinstance(obj, foast.TupleExpr):
        processed_elements = [process_elements(el, process_func, **kwargs) for el in obj.elts]
        return im.promote_to_lifted_stencil(lambda *elts: im.make_tuple(*elts))(
            *[self.visit(el, **kwargs) for el in processed_elements])
    else:
        return self._map(im.lambda_("it")(process_func("it")), obj)
```

Reply:
Great, this is exactly the pattern I meant. However this will fail in case the tuple is not constructed directly, but for example using a function call or in a previous assignment, e.g. this will not work:
```
tup = (foo, bar)
return `astype(tup, int)`
```
or
```
tup1 = (foo, bar)
tup2 = (baz, tup1)
return `astype(tup2, int)`
```
or
```
return `astype(field_op_returning_a_tuple(...), int)`
```
Please add test cases for the first two ones or adopt the existing ones (latter preferred to reduce code bloat).

Reply:
I adapted here below the pattern to also work for function calls and previous assignment and integrated it in the code.  The corresponding integration tests were also added. Let me know if the function is ok.
```
    def _process_elements(self, obj, process_func, **kwargs):
        """
        Recursively applies a processing function to the elements of a structured object, preserving its structure.
        Args:
            obj: The structured object to be processed.
            process_func: A function to apply to each element.
            **kwargs: Additional keyword arguments for the processing function.
        Example:
            result = process_elements(obj, lambda x: im.call("cast_")(x, str(dtype)), **kwargs)
        Returns:
            Structured object with the processing function applied to its elements.
        """
        if isinstance(obj.type, ts.TupleType):
            if isinstance(obj, foast.Name) or isinstance(obj, foast.Call):
                return self.visit(obj)
            processed_elements = [
                self._process_elements(el, process_func, **kwargs) for el in obj.elts
            ]
            return im.promote_to_lifted_stencil(lambda *elts: im.make_tuple(*elts))(
                *[el for el in processed_elements]
            )
        else:
            return self._map(im.lambda_("it")(process_func("it")), obj)
```

Reply:
I adapted the function in order to handle the cast of tuples and nested tuples that were not created directly like you mentioned above:
```
    def _process_elements(self, obj, obj_type, process_func, expr, **kwargs):
        """Recursively applies a processing function to the elements of a structured object, preserving its structure."""
        if isinstance(obj_type, ts.TupleType):
            if isinstance(obj, foast.Name) or isinstance(obj, foast.Call):
                return im.promote_to_lifted_stencil(lambda *elts: im.make_tuple(*elts))(
                    *[
                        self._process_elements(
                            obj, obj_type.types[i], process_func, im.tuple_get(i, expr)
                        )
                        for i in range(len(obj.type.types))
                    ]
                )
            elif isinstance(obj, foast.TupleExpr):
                return im.promote_to_lifted_stencil(lambda *elts: im.make_tuple(*elts))(
                    *[
                        self._process_elements(el, el.type, process_func, expr, **kwargs)
                        for el in obj.elts
                    ]
                )
        else:
            return self._map(im.lambda_("expr")(process_func(expr)), obj)
```
The tests were also adapted in the form here below to avoid that the backend silently does the cast when writing back the float output to an int field.
```
@gtx.field_operator
    def cast_tuple(
        a: cases.IFloatField, b: cases.IFloatField,
        a_casted_to_int_outside_of_gt4py: cases.IIntField,
        b_casted_to_int_outside_of_gt4py: cases.IIntField,
    ) -> tuple[gtx.Field[[IDim], bool], gtx.Field[[IDim], bool]]:
        result = astype(field_op_returning_a_tuple(a, b), int32)
        return result[0] == a_casted_to_int_outside_of_gt4py, result[1] == b_casted_to_int_outside_of_gt4py
```